source/_posts/vue-design-patterns.md

```diff
@@ -0,0 +1,988 @@
+        options: CookieOptions | undefined = undefined
+      ): string | undefined => {
+        const response = cookie.set(key, value, options);
+        update();
+        return response;
+      };
+    
+      const remove = (key: string) => {
+        cookie.remove(key);
+        update();
+      };
+    
+      return { cookies, get, set, remove };
+    }
+```
+
+The `ref()` method is a function that gives reactivity to a variable, `cookies`. In this case, when we talk about composables, we are directly talking about state and action exposure. This composable exposes three actions: `get()`, `set()`, and `remove()`; those do the same thing as the `CookiesAdapter` methods but are necessary to share through components as a composable, the same with `cookies` which is the general state of our Cookies. The trick behind the `update()` method synchronizes the cookies' state every time an action is called.
+
+Now, we can use it everywhere.
+
+```ts
+    // File: ComposablePatternView.vue
+    
+    <template>
+      <h4>Add Cookies</h4>
+      <input type="text" name="key" v-model="key" placeholder="Introduce a key" />
+      <input
+        type="text"
+        name="value"
+        v-model="value"
+        placeholder="Introduce a value"
+      />
+      <button @click="set(key, value)">Add/Update Cookie</button>
+      <br />
+    
+      <h4>Remove Cookies</h4>
+      <select name="cookies" v-model="selectedKey">
+        <option v-for="cookieKey in keys" :key="cookieKey" :value="cookieKey">
+          {{ cookieKey }}
+        </option>
+      </select>
+      <button @click="remove(selectedKey)">Remove Cookie</button>
+      <br />
+      <br />
+    
+      <pre>{{ cookies }}</pre>
+    </template>
+    
+    <script lang="ts">
+    import useCookies from "@/utils/useCookies";
+    import { computed, defineComponent, ref } from "vue";
+    
+    export default defineComponent({
+      setup() {
+        const { get, set, remove, cookies } = useCookies();
+        const selectedKey = ref(Object.keys(cookies.value)[0] || "");
+        const key = ref("");
+        const value = ref("");
+        const keys = computed(() => Object.keys(cookies.value));
+    
+        return { key, value, keys, selectedKey, get, set, remove, cookies };
+      },
+    });
+    </script>
+    
+    <style scoped>
+    input,
+    button {
+      margin: 5px;
+    }
+    </style>
+```
+
+Many people use this pattern without the reactive feature, but it is great to have.
+
+{% img "composable-example.png" "cookie editor with Vue, ui components (Composables pattern)" "lazy" %}
+
+This cookie composable is cool, but if you don't want to reinvent the wheel, the Vue ecosystem has many composition functions on the hand of [VueUse](https://vueuse.org/) and [vue-composable](https://pikax.me/vue-composable) libraries.
+
+> Code:  [vue design pattern repository](https://github.com/cr0wg4n/vue-design-patterns), execute, and go to **/composable-pattern** route 😎.
+
+
```

Comment:
There are several places with duplicated new lines.

- [ ] remove duplicated new lines across the entire article

Reply:
Done :D